### ISSUE TITLE
IClassListener is executed on each test method invocation rather than once per class

### ISSUE THREAD

**Issue by skirkpatrick** [cbeust/testng#921]
Listeners implementing `IClassListener` are running on each test method invocation, resulting in multiple calls per class.

I've created a simple Maven project demonstrating the issue (run with `mvn test`): https://github.com/skirkpatrick/TestNG-Class-Listener-Bug

Expected output:
```
Running TestSuite
onBeforeClass
test 1
test 2
onAfterClass
```

Actual output:
```
Running TestSuite
onBeforeClass
test 1
onAfterClass
onBeforeClass
test 2
onAfterClass
```

**Comment by juherr** [comment]
In fact, the expected output is:

```
onBeforeClass (test 1)
onBeforeClass (test 2)
test 1
test 2
onAfterClass (test 1)
onAfterClass (test 2)
```

But I'm agree with you, we should add 2 methods without `IMethodInstance` too.

**Comment by skirkpatrick** [comment]
Okay, that makes sense. Just for clarification, the methods without the IMethodInstance would behave like the no-argument `@BeforeClass` and `@AfterClass` methods? i.e., it would produce the output that I was originally expecting?

**Comment by juherr** [comment]
> the methods without the IMethodInstance would behave like the no-argument @BeforeClass and @AfterClass methods?

Yes! 

To be honest, the current behavior was not expected and you've just found an issue I didn't see: I didn't make enough test cases when I implemented the listener. And I was surprised to be notified many times too. 
But once you have a look at the parameters, it makes sense. 

As a workaround, I register notified classes into a collection and skip them after.

**Comment by juherr** [comment]
Fixed by #1047
